### ISSUE TITLE
Allow a user to enable/disable autorenewal of certs

### ISSUE THREAD

**Issue by ohemorange** [certbot/certbot#4504]
From #338.

This should probably be done via a subcommand or flag to add a field to the renewal config file, which is tested during `renew`.

**Comment by SwartzCr** [comment]
Here's a more full explanation from @zjs from that pr:

The "disable" operation would update the appropriate renewal configuration file with a property. When renew is identifying the certificates to renew, it would ignore/skip any which have this property.

From a user's point of view, this would allow them to:

Create several certificates.
Disable automatic renewal on one or more.
Put a call to certbot renew in their crontab to automatically renew all but those disabled in step 2.
This PR seems to do something else: it adds a flag to globally disable renewal. Do you have a use case in mind for this?

**Comment by SwartzCr** [comment]
@ohemorange do you feel that this is still needed? Or do you think users could manage this with a renewal hook?

**Comment by ohemorange** [comment]
I'm just the messenger here - I was just separating the previous PR into its component parts.

**Comment by r5d** [comment]
I would like to work on this issue.

I've made a rough patch: https://github.com/certbot/certbot/compare/master...r5d:4504-disable-enable-renew-cert-patch-02 (want to know if I'm going in the right direction).

The enable/disable workflow:

The renewal conf for (test) cert `ara.ricketyspace.net` looks like
this initially:
```
  # renew_before_expiry = 30 days
   version = 0.22.0.dev0
  archive_dir = /tmp/leitfJ7Z/conf/archive/ara.ricketyspace.net
  cert = /tmp/leitfJ7Z/conf/live/ara.ricketyspace.net/cert.pem
  privkey = /tmp/leitfJ7Z/conf/live/ara.ricketyspace.net/privkey.pem
  chain = /tmp/leitfJ7Z/conf/live/ara.ricketyspace.net/chain.pem
  fullchain = /tmp/leitfJ7Z/conf/live/ara.ricketyspace.net/fullchain.pem

  # Options used in the renewal process
  [renewalparams]
  account = d9302309ddf75295fc278b68b5df89b4
  work_dir = /tmp/leitfJ7Z/work
  webroot_path = /home/s/pub,
  server = https://acme-staging.api.letsencrypt.org/directory
  authenticator = webroot
  logs_dir = /tmp/leitfJ7Z/logs
  installer = None
  config_dir = /tmp/leitfJ7Z/conf
  [[webroot_map]]
  ara.ricketyspace.net = /home/s/pub
```

After executing:
```bash
  certbot renew --cert-name ara.ricketyspace.net --disable --config-dir /tmp/leitfJ7Z/conf --work-dir /tmp/leitfJ7Z/work --logs-dir /tmp/leitfJ7Z/logs
  Saving debug log to /tmp/leitfJ7Z/logs/letsencrypt.log

  -------------------------------------------------------------------------------
  Processing /tmp/leitfJ7Z/conf/renewal/ara.ricketyspace.net.conf
  -------------------------------------------------------------------------------

  -------------------------------------------------------------------------------
  Disabled renewal for ara.ricketyspace.net
  -------------------------------------------------------------------------------
```

The renewal conf file becomes:

```
  # renew_before_expiry = 30 days
  version = 0.22.0.dev0
  archive_dir = /tmp/leitfJ7Z/conf/archive/ara.ricketyspace.net
  cert = /tmp/leitfJ7Z/conf/live/ara.ricketyspace.net/cert.pem
  privkey = /tmp/leitfJ7Z/conf/live/ara.ricketyspace.net/privkey.pem
  chain = /tmp/leitfJ7Z/conf/live/ara.ricketyspace.net/chain.pem
  fullchain = /tmp/leitfJ7Z/conf/live/ara.ricketyspace.net/fullchain.pem

  # Options used in the renewal process
  [renewalparams]
  account = d9302309ddf75295fc278b68b5df89b4
  work_`dir = /tmp/leitfJ7Z/work
  server = https://acme-staging.api.letsencrypt.org/directory
  authenticator = webroot
  logs_dir = /tmp/leitfJ7Z/logs
  config_dir = /tmp/leitfJ7Z/conf
  disabled = True
  [[webroot_map]]
  ara.ricketyspace.net = /home/s/pub
```

When renewal is tried now:

```bash
  certbot renew --cert-name ara.ricketyspace.net --config-dir /tmp/leitfJ7Z/conf --work-dir /tmp/leitfJ7Z/work --logs-dir /tmp/leitfJ7Z/logs
  Saving debug log to /tmp/leitfJ7Z/logs/letsencrypt.log

  -------------------------------------------------------------------------------
  Processing /tmp/leitfJ7Z/conf/renewal/ara.ricketyspace.net.conf
  -------------------------------------------------------------------------------
  Cert renewal is disabled
                                                                                                                                                                                                                      -------------------------------------------------------------------------------                                                                                                                                                                                                                                                                                                                                                         The following certs are not due for renewal yet:
    /tmp/leitfJ7Z/conf/live/ara.ricketyspace.net/fullchain.pem (skipped)
  No renewals were attempted.
  -------------------------------------------------------------------------------
```
After executing:

```bash
  certbot renew --cert-name ara.ricketyspace.net --enable --config-dir /tmp/leitfJ7Z/conf --work-dir /tmp/leitfJ7Z/work --logs-dir /tmp/leitfJ7Z/logs
  Saving debug log to /tmp/leitfJ7Z/logs/letsencrypt.log
  -------------------------------------------------------------------------------
  Processing /tmp/leitfJ7Z/conf/renewal/ara.ricketyspace.net.conf
  -------------------------------------------------------------------------------

  -------------------------------------------------------------------------------
  Enabled renewal for ara.ricketyspace.net
  -------------------------------------------------------------------------------
```

renewal conf for `ara.ricketyspace` updates to:

```
  # renew_before_expiry = 30 days
  version = 0.22.0.dev0
  archive_dir = /tmp/leitfJ7Z/conf/archive/ara.ricketyspace.net
  cert = /tmp/leitfJ7Z/conf/live/ara.ricketyspace.net/cert.pem
  privkey = /tmp/leitfJ7Z/conf/live/ara.ricketyspace.net/privkey.pem
  chain = /tmp/leitfJ7Z/conf/live/ara.ricketyspace.net/chain.pem
  fullchain = /tmp/leitfJ7Z/conf/live/ara.ricketyspace.net/fullchain.pem

  # Options used in the renewal process
  [renewalparams]
  account = d9302309ddf75295fc278b68b5df89b4
  work_dir = /tmp/leitfJ7Z/work
  server = https://acme-staging.api.letsencrypt.org/directory
  authenticator = webroot
  logs_dir = /tmp/leitfJ7Z/logs
  config_dir = /tmp/leitfJ7Z/conf
  disabled = False
  [[webroot_map]]
  ara.ricketyspace.net = /home/s/pub
```

If renewal is tried now:

```bash
  certbot renew --cert-name ara.ricketyspace.net --config-dir /tmp/leitfJ7Z/conf --work-dir /tmp/leitfJ7Z/work --logs-dir /tmp/leitfJ7Z/logs
  Saving debug log to /tmp/leitfJ7Z/logs/letsencrypt.log

  -------------------------------------------------------------------------------
  Processing /tmp/leitfJ7Z/conf/renewal/ara.ricketyspace.net.conf
  -------------------------------------------------------------------------------
  Cert not yet due for renewal

  -------------------------------------------------------------------------------

  The following certs are not due for renewal yet:
    /tmp/leitfJ7Z/conf/live/ara.ricketyspace.net/fullchain.pem (skipped)
  No renewals were attempted.
  -------------------------------------------------------------------------------

```

  Observations:

  * The `webroot_path` and `installer` fields are getting removed from
     `renewalparams` (in renewal conf file) after the first `certbot
     renew --cert-name ... --disable` or `certbot renew --cert-name
     ... --enable`

     So, I think the way I'm saving renewal conf file is wrong
     ([renewal.py:417](https://github.com/r5d/certbot/blob/4504-disable-enable-renew-cert-patch-01/certbot/renewal.py#L418)). What is the correct way to update the renewal
     config?

**Comment by ohemorange** [comment]
Hi @rd5, excited that you're interested in working on this (unless you're no longer able to?).

We already have a defunct `autorenew` param that if set to `False` would turn off autorenewing. It probably just needs to be added to `cli.py`, `constants.py`, and `renewal.py`; if you `git grep http01_address` or `must_staple` you'll see examples of where it should go. Then it should theoretically just work with the logic already in there? Would love to have you try that out!

**Comment by r5d** [comment]
@ohemorange, I've created another rough patch based on your feedback. It would be great if you can check if I'm going in the right direction: https://github.com/certbot/certbot/compare/master...r5d:4504-renew-autorenew-option-02

In this patch, the `autorenew` is checked for and the renewal config is [updated in the `certbot.renewal.handle_renewal_request` function](https://github.com/certbot/certbot/compare/master...r5d:4504-renew-autorenew-option-02#diff-0de5bd9f8d55f4cb662d41832cbd9b01R411); Is this the correct place to do this or should it be done elsewhere?

Also, may I open PR for this?

**Comment by strugee** [comment]
@r5d you should probably go ahead and open a PR. That's what they're for - to get feedback and reviews.

**Comment by ohemorange** [comment]
Definitely go ahead and open a PR, looks like a good start! Review preview: going to request that instead of storing as 0 or 1 as it currently does, to switch to storing booleans as we do elsewhere. Plus, the default should be places in `constants.py`, and it might be cleaner to use `--autorenew` and `--no-autorenew` to match our existing naming patterns.

**Comment by r5d** [comment]
@ohemorange, open a [PR](#5911).

**Comment by schoen** [comment]
The issue that I'm puzzled by (that I'm not sure @r5d's patches address yet) is why `autorenew = False` doesn't actually inhibit autorenewal in the current version Certbot! (I think that's why @ohemorange referred to this feature as "defunct".)

**Comment by bmw** [comment]
Fixed in #5911.
